core/src/com/unciv/models/ruleset/Unique.kt

```diff
@@ -14,16 +14,76 @@ import com.unciv.models.translations.hasPlaceholderParameters
 import com.unciv.ui.worldscreen.unit.UnitActions
 import kotlin.random.Random
 
-class Unique(val text:String){
+
+// parameterName values should be compliant with autogenerated values in TranslationFileWriter.generateStringsFromJSONs
+// Eventually we'll merge the translation generation to take this as the source of that
+enum class UniqueParameterType(val parameterName:String, val complianceCheck:(String, Ruleset)->Boolean) {
+    Number("amount", { s, r -> s.toIntOrNull() != null }),
+    UnitFilter("unitType", { s, r -> r.unitTypes.containsKey(s) || unitTypeStrings.contains(s) }),
+    Unknown("",{s,r -> true});
+
+    companion object {
+        val unitTypeStrings = hashSetOf(
+            "Military",
+            "Civilian",
+            "non-air",
+            "relevant",
+            "Nuclear Weapon",
+            "City",
+            // These are up for debate
+            "Air",
+            "land units",
+            "water units",
+            "air units",
+            "military units",
+            "submarine units",
+            // Note: this can't handle combinations of parameters (e.g. [{Military} {Water}])
+        )
+    }
+}
+
+enum class UniqueType(val text:String){
+
+    ConsumesResources("Consumes [amount] [resource]");
+
+    /** For uniques that have "special" parameters that can accept multiple types, we can override them manually
+     *  For 95% of cases, auto-matching is fine. */
+    val parameterTypeMap= ArrayList<List<UniqueParameterType>>()
+    init {
+        for (placeholder in text.getPlaceholderParameters()) {
+            val matchingParameterType =
+                UniqueParameterType.values().firstOrNull { it.parameterName == placeholder }
+            if (matchingParameterType != null) parameterTypeMap.add(listOf(matchingParameterType))
+            else parameterTypeMap.add(listOf(UniqueParameterType.Unknown))
+        }
+    }
+    val placeholderText = text.getPlaceholderText()
+
+    fun checkCompliance(unique:Unique, ruleset: Ruleset):Boolean {
+        for ((index, param) in unique.params.withIndex())
+            if (parameterTypeMap[index].none { it.complianceCheck(param, ruleset) })
+                return false
+        return true
+    }
+}
+
+
+
+class Unique(val text:String) {
     val placeholderText = text.getPlaceholderText()
     val params = text.getPlaceholderParameters()
+
     /** This is so the heavy regex-based parsing is only activated once per unique, instead of every time it's called
      *  - for instance, in the city screen, we call every tile unique for every tile, which can lead to ANRs */
     val stats: Stats by lazy {
         val firstStatParam = params.firstOrNull { Stats.isStats(it) }
         if (firstStatParam == null) Stats() // So badly-defined stats don't crash the entire game
         else Stats.parse(firstStatParam)
     }
+
+    fun isOfType(uniqueType: UniqueType) = placeholderText == uniqueType.placeholderText
```

Comment:
Could be performance-critical. (loud thinking) Can we avoid the char-by-char equality? Intern both uniqueType.placeholderText and Unique.placeholderText once and early and then use `===` here? Or cache the UniqueType per Unique? Both won't be good if Unique objects are short lived - are they?